code/StepExporter.cpp

```diff
@@ -0,0 +1,380 @@
+	outfile->Write( iDoTheExportThing.mOutput.str().c_str(), static_cast<size_t>(iDoTheExportThing.mOutput.tellp()),1);
+}
+
+} // end of namespace Assimp
+
+
+namespace {
+inline uint64_t toIndexHash(int32_t id1, int32_t id2)
+{
+	// dont wonder that -1/-1 -> hash=-1
+    uint64_t hash = (uint32_t) id1;
+    hash = (hash << 32);
+    hash += (uint32_t) id2;
+    return hash;
+}
+
+inline void fromIndexHash(uint64_t hash, int32_t &id1, int32_t &id2)
+{
+    id1 = (hash & 0xFFFFFFFF00000000) >> 32;
+    id2 = (hash & 0xFFFFFFFF);
+}
+
+// Collect world transformations for each node
+	void CollectTrafos(const aiNode* node, std::map<const aiNode*, aiMatrix4x4>& trafos) {
+		const aiMatrix4x4& parent = node->mParent ? trafos[node->mParent] : aiMatrix4x4();
+		trafos[node] = parent * node->mTransformation;
+		for (unsigned int i = 0; i < node->mNumChildren; ++i) {
+			CollectTrafos(node->mChildren[i], trafos);
+		}
+	}
+
+	// Generate a flat list of the meshes (by index) assigned to each node
+	void CollectMeshes(const aiNode* node, std::multimap<const aiNode*, unsigned int>& meshes) {
+		for (unsigned int i = 0; i < node->mNumMeshes; ++i) {
+			meshes.insert(std::make_pair(node, node->mMeshes[i]));
+		}
+		for (unsigned int i = 0; i < node->mNumChildren; ++i) {
+			CollectMeshes(node->mChildren[i], meshes);
+		}
+	}
+}
+
+// ------------------------------------------------------------------------------------------------
+// Constructor for a specific scene to export
+StepExporter::StepExporter(const aiScene* pScene, IOSystem* pIOSystem, const std::string& path, const std::string& file, const ExportProperties* pProperties) : mIOSystem(pIOSystem), mPath(path), mFile(file), mProperties(pProperties)
+{
+	CollectTrafos(pScene->mRootNode, trafos);
+	CollectMeshes(pScene->mRootNode, meshes);
+
+	// make sure that all formatting happens using the standard, C locale and not the user's current locale
+	mOutput.imbue( std::locale("C") );
+
+	mScene = pScene;
+	mSceneOwned = false;
+
+	// set up strings
+	endstr = ";\n"; 
+
+	// start writing
+	WriteFile();
+
+
+
+}
+
+// ------------------------------------------------------------------------------------------------
+// Destructor
+StepExporter::~StepExporter()
+{
+	if(mSceneOwned) {
+		delete mScene;
+	}
+}
+
+// ------------------------------------------------------------------------------------------------
+// Starts writing the contents
+void StepExporter::WriteFile()
+{
+	// see http://shodhganga.inflibnet.ac.in:8080/jspui/bitstream/10603/14116/11/11_chapter%203.pdf
+	// note, that all realnumber values must be comma separated in x files
+	mOutput.setf(std::ios::fixed);
+	mOutput.precision(16); // precission for double
+	
+	int ind = 100; // the start index to be used
+	int faceEntryLen = 30; // number of entries for a triangle/face
+	// prepare unique (count triangles and vertices)
+	std::map<aiVector3D*, int> uniqueVerts; // use a map to reduce find complexity to log(n)
```

Comment:
I think a hashmap is the preferred data structure here. Ordering doesn't matter and it'll give you amortized O(1).

Have a look at `code/FBXCompileConfig.h` for how to get a safe `unordered_map` type.